File: leetcode/python/0125. Valid Palindrome.py
# Time complexity: O(N)
# Space complexity: O(N)

class Solution(object):
    def isPalindrome(self, s):
        lower = s.lower()

        a = ord("a")
        z = ord("z")
        digit_0 = ord("0")
        digit_9 = ord("9")

        str_list = []
        
        for char in lower:
            curr = ord(char)
            if curr >= a and curr <= z or curr >= digit_0 and curr <= digit_9:
                str_list.append(char)
        
        left = 0
        right = len(str_list) - 1

        while right >= left:
            if str_list[left] != str_list[right]:
                return False
            left += 1
            right -= 1
        return True 

File: leetcode/java/0125. Valid Palindrome.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public boolean isPalindrome(String s) {
        StringBuilder sb = new StringBuilder();

        for (char c : s.toCharArray()) {
            if (Character.isLetter(c) || Character.isDigit(c)) {
                sb.append(Character.toLowerCase(c));
            }
        }

        int left = 0;
        int right = sb.length() - 1;

        while (right >= left) {
            if (sb.charAt(right) == sb.charAt(left)) {
                left += 1;
                right -= 1;
            } else {
                return false;
            }
        }
        return true;
    }
}